fs/shm/shmfs.c

```diff
@@ -0,0 +1,384 @@
+    {
+      delete_shm_object(inode->i_private);
+      inode->i_private = NULL;
+    }
+
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: shmfs_map_object
+ ****************************************************************************/
+
+static int shmfs_map_object(FAR struct shmfs_object_s *object,
+                            FAR void **vaddr)
+{
+  int ret = OK;
+
+#ifdef CONFIG_BUILD_KERNEL
+  /* Map the physical pages of the shm object with MMU. */
+
+  FAR struct tcb_s        *tcb   = nxsched_self();
+  FAR struct task_group_s *group = tcb->group;
+  FAR uintptr_t *pages =
+    (FAR uintptr_t *)&object->paddr;
+  uintptr_t                mapaddr;
+  unsigned int             npages;
+
+  /* Find a free vaddr space that satisfies length */
+
+  mapaddr = (uintptr_t)vm_alloc_region(get_group_mm(group), 0,
+                                       object->length);
+  if (mapaddr == 0)
+    {
+      return -ENOMEM;
+    }
+
+  /* Convert the region size to pages */
+
+  npages = MM_NPAGES(object->length);
+
+  /* Map the memory to user virtual address space */
+
+  ret = up_shmat(pages, npages, mapaddr);
+  if (ret < 0)
+    {
+      vm_release_region(get_group_mm(group), (FAR void *)mapaddr,
+                        object->length);
+    }
+  else
+    {
+      *vaddr = (FAR void *)mapaddr;
+    }
+#else
+  /* Use the physical address directly */
+
+  *vaddr = object->paddr;
+#endif
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_mmap
+ ****************************************************************************/
+
+static int shmfs_mmap(FAR struct file *filep,
+                      FAR struct mm_map_entry_s *entry)
+{
+  FAR struct shmfs_object_s *object;
+  int ret = -EINVAL;
+
+  DEBUGASSERT(filep->f_inode != NULL);
+
+  /* We don't support offset at the moment, just mapping the whole object */
+
+  if (entry->offset != 0)
+    {
+      return ret;
+    }
+
+  /* Keep the inode when mmapped, increase refcount */
+
+  ret = inode_addref(filep->f_inode);
+  if (ret >= 0)
+    {
+      object = (FAR struct shmfs_object_s *)filep->f_inode->i_private;
+
+      if (object)
+        {
+          ret = shmfs_map_object(object, &entry->vaddr);
+        }
+
+      if (!object || ret < 0)
+        {
+          /* Oops, the file has not been truncated yet or
+           * there was error in up_shmat
+           */
+
+          inode_release(filep->f_inode);
+        }
+      else
+        {
+          entry->munmap = shmfs_munmap;
+          entry->priv.p = (FAR void *)filep->f_inode;
+          mm_map_add(entry);
+        }
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: shmfs_unmap_object
+ ****************************************************************************/
+
+static int shmfs_unmap_area(FAR struct task_group_s *group,
```

Comment:
merge into shmfs_munmap?

Reply:
I liked to have this separate since it is CONFIG_BUILD_KERNEL specific, so it was nicer to have it in it's own function

Reply:
I did this part and I prefer to have the static method here. It abstracts the build type quite nicely IMO.

Reply:
Ok